[ONNX] Support tensorList when Infers shape and type uninitialized output

hwangdeyu opened the pull request:
Fixes #59179

Error message:
>ONNX export: RuntimeError: rINTERNAL ASSERT FAILED at "..\\aten\\src\\ATen/core/jit_type_base.h":172

Cause tensor list is not support when Infers shape and type uninitialized output, so add it.

hwangdeyu committed:
add tensorList unintialized case

hwangdeyu committed:
fix format

hwangdeyu committed:
add sequance loopcarried test

BowenBao commented on torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:
```diff
@@ -258,25 +259,41 @@ bool IsUninitializedNode(Node* n) {
-    const_node->t_(attr::value, at::zeros({}, elem_type));
-    const_node->output()->setType(
-        TensorType::create(*(output_type->scalarType()), at::kCPU, {}, {}));
+  Node* const_node;
```
Try to avoid uninitialized variable.

hwangdeyu replied:
Thanks, initialize it to nullptr.

BowenBao commented on torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:
```diff
@@ -258,25 +259,41 @@ bool IsUninitializedNode(Node* n) {
-    const_node->output()->setType(
-        TensorType::create(*(output_type->scalarType()), at::kCPU, {}, {}));
+  Node* const_node;
+  if (other_output->type()->cast<TensorType>() != nullptr) {
```
```suggestion
  if (auto other_type = other_output->type()->cast<TensorType>()) {
```

BowenBao commented on torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:
```diff
@@ -258,25 +259,41 @@ bool IsUninitializedNode(Node* n) {
+      const_node->output()->setType(
+          TensorType::create(*(output_type->scalarType()), at::kCPU, {}, {}));
+    }
+  } else if (other_output->type()->cast<ListType>() != nullptr) {
```
```suggestion
  } else if (auto output_type = other_output->type()->cast<ListType>()) {
```

BowenBao commented on torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:
```diff
@@ -258,25 +259,41 @@ bool IsUninitializedNode(Node* n) {
+    auto output_type = other_output->type()->expect<ListType>();
+    TypePtr elem = output_type->getElementType();
+    const_node = graph->create(::c10::onnx::SequenceEmpty, 1);
+    if (elem->cast<TensorType>()->scalarType().has_value()) {
```
condition need to check that `elem->cast<TensorType>()` is not nullptr

BowenBao commented on torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:
```diff
@@ -258,25 +259,41 @@ bool IsUninitializedNode(Node* n) {
+      auto scalar_type = elem->cast<TensorType>()->scalarType().value();
+      auto onnx_type = ATenTypeToOnnxType(scalar_type);
+      const_node->i_(attr::dtype, onnx_type);
+      const_node->output()->setType(ListType::create(elem->cast<TensorType>()));
```
this will copy over tensor shape from elem as well. is this intended?

hwangdeyu replied:
I thought that we copy over tensor shape from elem. I'm not particularly sure if that makes sense.
Is that more reasonable to use `setType(other_output->type())`?

BowenBao replied:
The current code is the same as `setType(other_output->type())`. `type` contains both `shape` and `scalarType`. If you intend to copy over both, then use `setType`, if you want to only copy `scalarType`, then you will need to update L#292

BowenBao commented on torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:
```diff
@@ -258,25 +259,41 @@ bool IsUninitializedNode(Node* n) {
+      const_node->i_(attr::dtype, onnx_type);
+      const_node->output()->setType(ListType::create(elem->cast<TensorType>()));
+    } else {
+      const_node->output()->setType(other_output->type());
```
so no scalar type to copy over from. this could be a problem, any thoughts on what we can do in general?

neginraoof replied:
@hwangdeyu Is this comment addressed?

hwangdeyu committed:
fix comments

hwangdeyu committed:
add comment for no scalar type in tensorList

jiafatom commented on torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:
```diff
@@ -258,25 +259,41 @@ bool IsUninitializedNode(Node* n) {
+          TensorType::create(*(output_type->scalarType()), at::kCPU, {}, {}));
+    }
+  } else if (auto output_type = other_output->type()->cast<ListType>()) {
+    TypePtr elem = output_type->getElementType();
```
For ListType case, I wonder what is the shape inference result look like. I am converting a silero model which uses this technique, looks like it returns Tensor[] but there is no shape for this Tensor[]. Seems that we need have a shape inference for each element in Tensor[].

jiafatom commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -6262,6 +6262,52 @@ def forward(self, y):
+                        y = y + 4
+                    else:
+                        return [y]
+                return [y]
```
Can we have a unit test, if branch returns tensor list of empty list, else branch returns tensor list of several elements?

jiafatom commented on torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:
```diff
@@ -258,25 +259,41 @@ bool IsUninitializedNode(Node* n) {
+    }
+  } else if (auto output_type = other_output->type()->cast<ListType>()) {
+    TypePtr elem = output_type->getElementType();
+    const_node = graph->create(::c10::onnx::SequenceEmpty, 1);
```
Where is the code you add Identity node? Once you insert a new node to graph, you may need add shape inference for that node. See https://github.com/pytorch/pytorch/blob/dec5aa2260cef540b622bd9a9504b6f11cb1f607/torch/csrc/jit/passes/onnx.cpp#L286

hwangdeyu replied:
thanks David！
I'm not sure if SequenceEmpty needs an certain shape, will check in ONNX spec and try to use this function.

hwangdeyu replied:
btw, add Identity node in thins link [fixup_onnx_controlflow.cpp#L190]. We only deal with the TensorList case when Infer Shape and Type For UninitializedOutput.(https://github.com/pytorch/pytorch/blob/e523674c51072036a69125f098eb16c93f32b59c/torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp#L190)

jiafatom committed:
Enhance shape (#60335)

Enhanced shape inference by introducing typeReliableMap.

Co-authored-by: jiafatom <jiafa@microsoft.com>

neginraoof commented on torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:
```diff
@@ -258,25 +259,43 @@ bool IsUninitializedNode(Node* n) {
+      auto onnx_type = ATenTypeToOnnxType(scalar_type);
+      const_node->i_(attr::dtype, onnx_type);
+      const_node->output()->setType(other_output->type());
+    } else {
```
I think we can also handle scalar type lists like (int[], float[], etc.) here.
Maybe like:
else if (elem->cast<IntType>()) {
      auto scalar_type = at::kLong;
      auto onnx_type = ATenTypeToOnnxType(scalar_type);
      const_node->i_(attr::dtype, onnx_type);
      const_node->output()->setType(other_output->type());
}

hwangdeyu replied:
Good point! Will add this part and tests.
By the way, Do you think we also need to add scalar type like(int, float etc..) in output_type?
https://github.com/pytorch/pytorch/blob/c3eb80300af37aa41b9e7ff2d3ff2220393db6b7/torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp#L284

hwangdeyu replied:
It seems the ONNX node type must be Tensor type. Even we return like the `intList`, it will also return the element tensor type. So we may not need to handle those type.
I have added this in test case.

hwangdeyu committed:
[ONNX] Remove aten parameter (#61652)

* remove aten argument in export()

* add export_to_pretty_string default value OperatorExportTypes.ONNX

* add DPYTORCH_ONNX_CAFFE2_BUNDLE description

Co-authored-by: Gary Miguel <garymiguel@microsoft.com>

BowenBao committed:
[ONNX] Fix controlflow shape inference with contrib op (#60707)

`ONNXShapeTypeInference` for node `n` is skipped if `n` is non ONNX namespace, or if `n` contains any non ONNX namespace nodes. This prevents controlflow nodes containing contrib ops from running `SpecialPostProcess`, which sets up correct node output shape/type information in rare cases.

This PR depends on opset 14 export https://github.com/pytorch/pytorch/pull/59486

Co-authored-by: BowenBao <bowbao@microsoft.com>

hwangdeyu committed:
add intList type test

fatcat-z committed:
[ONNX] Deprecate enable_onnx_checker argument in torch.onnx.export() (#61708)

As of now, the "enable_onnx_checker" parameter was described as below:

enable_onnx_checker (bool, default True): If True the ONNX model checker will be run to ensure the exported model is a valid ONNX model.

An invalid ONNX graph is useless to users so such checker should be done for each call.

In this PR, we will still write the model to an ONNX file even it is invalid. And the exception will be thrown after the ONNX file has been created. This enables user output an invalid ONNX graph for debug.

This PR will still keep it in torch.onnx.export() function for backward support while all backend logic has been changed to work as enable_onnx_checker is set to True.

Co-authored-by: fatcat-z <zhang-ji@outlook.com>

hwangdeyu committed:
fix flake8

jiafatom commented on torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:
```diff
@@ -256,25 +257,43 @@ bool IsUninitializedNode(Node* n) {
 // output of the other subblock. prim::Uninitialized node is proven to be
-// unused. So replace this node with a constant of the inferred shape and type.
+// unused. So replace this node with a constant(TensorType) or
+// Sequence(ListType) of the inferred shape and type.
```
nit: add space

hwangdeyu commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -6387,6 +6387,92 @@ def forward(self, y):
                       input_names=["input_1"],
                       dynamic_axes={"input_1": [0, 1]})
 
+    # ONNX::identity of sequence supported for ONNX opset >= 14
```
Thanks, using onnx::Identity makes it more precise.

hwangdeyu committed:
fix test error and word format

hwangdeyu commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -6518,6 +6518,92 @@ def forward(self, y):
+                      dynamic_axes={"input_1": [0, 1]})
+
+    # onnx::Identity of sequence supported for ONNX opset >= 14
+    # ORT is not support this graph in version 14.
```
Need check if the graph is vaild in ONNX

hwangdeyu replied:
It's valid.

hwangdeyu commented:
@neginraoof  https://github.com/pytorch/pytorch/pull/60534#discussion_r701456509
Yes, I have added this warning information.

jiafatom commented on torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:
```diff
@@ -256,25 +257,43 @@ bool IsUninitializedNode(Node* n) {
+      const_node->output()->setType(other_output->type());
+    }
   }
   const_node->insertBefore(block->return_node());
```
Probably we need add shape inference for this const node, like
```cpp
const ParamMap empty_params_dict = {};
auto opset_version =
        py::cast<int>(onnx_symbolic.attr("_export_onnx_opset_version"));
ONNXShapeTypeInference(const_node, empty_params_dict, opset_version);
```
https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/onnx.cpp#L286

hwangdeyu replied:
> Probably we need add shape inference for this const node, like
> 
> ```c++
> const ParamMap empty_params_dict = {};
> auto opset_version =
>         py::cast<int>(onnx_symbolic.attr("_export_onnx_opset_version"));
> ONNXShapeTypeInference(const_node, empty_params_dict, opset_version);
> ```
> 
> https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/onnx.cpp#L286

Done, shape inference needed after `constant` node.

jiafatom replied:
nit: It is better to move `const_node->insertBefore(block->return_node());` before shape inference logic. (although for const node it may be fine)

hwangdeyu committed:
add shape inference check for const_node

jiafatom reviewed (approved):
Nice job! Thanks!

BowenBao commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -6518,6 +6518,92 @@ def forward(self, y):
+    # onnx::Identity of sequence supported for ONNX opset >= 14
+    @skipIfUnsupportedMinOpsetVersion(14)
+    @skipIfONNXShapeInference(False)
+    def test_uninitialized_intList(self):
```
hi @hwangdeyu , this test starts to fail after syncing with latest master + onnx_ms_1 branch. Please take a look when you get a chance. I'll submit the PR to pytorch master with this test disabled.

jiafatom replied:
filed a work item: https://msdata.visualstudio.com/Vienna/_workitems/edit/1441852

BowenBao replied:
update: we discovered it is due to upstream torch IR graph changes that led to the regression. The initial int list creation is noe emitted as `prim::ListConstruct` instead of `prim::Loop`.
Test case is updated to preserve the previous expected IR graph. The item above can be used to track future support for the new IR graph.
